### PR TITLE
[RLlib] Integration with SUMO Simulator

### PR DESCRIPTION
**Pull request by lcodeca**
## Why are these changes needed?

Eclipse SUMO (Simulation of Urban MObility) is an open-source, highly portable, microscopic, and continuous multi-modal traffic simulation package designed to handle large networks. See: https://www.eclipse.org/sumo/

This PR integrates the RLLIB SUMO Utils library (https://github.com/lcodeca/rllibsumoutils) directly in RLLIB.
It provides: 
- `rllib/contrib/sumo` a generic connector between a user-defined environment and the SUMO simulator with additional utils.
- `rllib/examples/simulators/sumo` and example of the multi-agent environment and a simple SUMO scenario
- `rllib/examples/sumo_env_local.py` PPO Trainer (tune)

The integration was discussed with @sven1977 

## Related issue number
None.

## Checks

- [] I've run `scripts/format.sh` to lint the changes in this PR.
Almost:
```
WARNING: Ray uses flake8 3.7.7, You currently are using 3.8.4. This might generate different results.
WARNING: Ray uses yapf 0.23.0, You currently are using 0.30.0. This might generate different results.
WARNING: Ray uses shellcheck 0.7.1, You currently are using 0.5.0. This might generate different results.
WARNING: Ray uses mypy 0.782, You currently are using 0.790. This might generate different results.
WARNING: Ray uses clang-format 7.0.0, You currently are using 9.0.0. This might generate different results.
From https://github.com/ray-project/ray
 * branch                master     -> FETCH_HEAD
Traceback (most recent call last):
  File "ci/travis/check_import_order.py", line 68, in <module>
    all_py_files = glob.glob("*.py", recursive=True)
TypeError: glob() got an unexpected keyword argument 'recursive'
```
Anyway, the `pylint --rcfile pylintrc` run was a 10/10

- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(
   - [x] Given that it requires the installation of SUMO simulator, I tested it in https://github.com/lcodeca/rllibsumodocker/tree/rllib-integration

### CODE REVIEW

**Comment by richardliaw** [comment]
cc @eugenevinitsky can you take a look at this :)

**Review by sven1977 — APPROVED** [review]
This is a great addition to RLlib. Thanks @lcodeca !

File: rllib/contrib/sumo/connector.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    # Anything. User defined.
+    "misc": None,
+}
+
+###############################################################################
+
+
+class SUMOConnector(object):
+    """ Handler of a SUMO simulation. """
+    def __init__(self, config):
```
Comment (sven1977) [inline rllib/contrib/sumo/connector.py:66]:
```suggestion

    def __init__(self, config):
```

File: rllib/contrib/sumo/connector.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+            self._sumo_parameters.extend(["--start", "--quit-on-end"])
+        if config["sumo_params"] is not None:
+            self._sumo_parameters.extend(config["sumo_params"])
+        self._sumo_parameters.extend(
+            ["--output-prefix", self._sumo_output_prefix])
+        logger.debug("SUMO command line: %s", str(self._sumo_parameters))
+        if config["trace_file"]:
+            traci.start(self._sumo_parameters,
+                        traceFile="{}.tracefile.log".format(
+                            self._sumo_output_prefix))
```
Comment (sven1977) [inline rllib/contrib/sumo/connector.py:107]:
```suggestion
            traci.start(
                self._sumo_parameters,
                traceFile="{}.tracefile.log".format(self._sumo_output_prefix))
```

File: rllib/contrib/sumo/utils.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+
+###############################################################################
+
+
+class SUMOUtils(SUMOConnector):
+    """
+        A wrapper for the interaction with the SUMO simulation that adds
+        functionalities.
+    """
+    def _initialize_metrics(self):
```
Comment (sven1977) [inline rllib/contrib/sumo/utils.py:51]:
```suggestion

    def _initialize_metrics(self):
```

File: rllib/examples/simulators/sumo/marlenvironment.py [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+
+
+###############################################################################
+
+MS_TO_KMH = 3.6
+
+
+class SUMOSimulationWrapper(SUMOUtils):
+    """ A wrapper for the interaction with the SUMO simulation """
+    def _initialize_simulation(self):
```
Comment (sven1977) [inline rllib/examples/simulators/sumo/marlenvironment.py:52]:
```suggestion

    def _initialize_simulation(self):
```

File: rllib/examples/simulators/sumo/marlenvironment.py [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+            self.end_simulation()
+        return True
+
+
+###############################################################################
+
+
+class SUMOAgent(object):
+    """ Agent implementation. """
+    def __init__(self, agent, config):
```
Comment (sven1977) [inline rllib/examples/simulators/sumo/marlenvironment.py:99]:
```suggestion
class SUMOAgent:
    """ Agent implementation. """

    def __init__(self, agent, config):
```

File: rllib/examples/simulators/sumo/marlenvironment.py [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+        sumo_handler.traci_handler.route.add(route, ["road"])
+        # insert the agent in the simulation
+        # traci.vehicle.add(self, vehID, routeID, typeID="DEFAULT_VEHTYPE",
+        #   depart=None, departLane="first", departPos="base", departSpeed="0",
+        #   arrivalLane="current", arrivalPos="max", arrivalSpeed="current",
+        #   fromTaz="", toTaz="", line="", personCapacity=0, personNumber=0)
+        sumo_handler.traci_handler.vehicle.add(self.agent_id,
+                                               route,
+                                               departLane="best",
+                                               departSpeed="max")
```
Comment (sven1977) [inline rllib/examples/simulators/sumo/marlenvironment.py:138]:
```suggestion
        sumo_handler.traci_handler.vehicle.add(
            self.agent_id, route, departLane="best", departSpeed="max")
```

File: rllib/examples/simulators/sumo/marlenvironment.py [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+        sumo_handler.traci_handler.vehicle.subscribeLeader(self.agent_id)
+        sumo_handler.traci_handler.vehicle.subscribe(self.agent_id,
+                                                     varIDs=[
+                                                         tc.VAR_SPEED,
+                                                     ])
```
Comment (sven1977) [inline rllib/examples/simulators/sumo/marlenvironment.py:143]:
```suggestion
        sumo_handler.traci_handler.vehicle.subscribe(
            self.agent_id, varIDs=[tc.VAR_SPEED])
```

File: rllib/examples/simulators/sumo/marlenvironment.py [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+    },
+    "max_speed": 130,  # km/h
+}
+
+
+class SUMOTestMultiAgentEnv(MultiAgentEnv):
+    """
+    A RLLIB environment for testing MARL environments with SUMO simulations.
+    """
+    def __init__(self, config):
```
Comment (sven1977) [inline rllib/examples/simulators/sumo/marlenvironment.py:177]:
```suggestion

    def __init__(self, config):
```

File: rllib/examples/simulators/sumo/marlenvironment.py [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+            self.rndgen.shuffle(shuffled_agents)  # in-place shuffle
+
+        # Take action
+        for agent in shuffled_agents:
+            self.agents[agent].step(action_dict[agent], self.simulation)
+
+        logger.debug("Before SUMO")
+        ongoing_simulation = self.simulation.step(until_end=False,
+                                                  agents=set(
+                                                      action_dict.keys()))
```
Comment (sven1977) [inline rllib/examples/simulators/sumo/marlenvironment.py:347]:
```suggestion
        ongoing_simulation = self.simulation.step(
            until_end=False, agents=set(action_dict.keys()))
```

File: rllib/examples/sumo_env_local.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+logging.basicConfig(level=logging.WARN)
+logger = logging.getLogger("ppotrain")
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--sumo-connect-lib",
+                    type=str,
+                    default="libsumo",
+                    choices=["libsumo", "traci"],
+                    help="The SUMO connector to import. "
+                    "Requires the env variable SUMO_HOME set.")
```
Comment (sven1977) [inline rllib/examples/sumo_env_local.py:35]:
```suggestion
parser.add_argument(
    "--sumo-connect-lib",
    type=str,
    default="libsumo",
    choices=["libsumo", "traci"],
    help="The SUMO connector to import. Requires the env variable SUMO_HOME set.")
```

File: rllib/examples/sumo_env_local.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+parser.add_argument(
+    "--sumo-gui",
+    action="store_true",
+    help="Enables the SUMO GUI. Possible only with TraCI connector.")
+parser.add_argument("--sumo-config-file",
+                    type=str,
+                    default=None,
+                    help="The SUMO configuration file for the scenario.")
```
Comment (sven1977) [inline rllib/examples/sumo_env_local.py:43]:
```suggestion
parser.add_argument(
    "--sumo-config-file",
    type=str,
    default=None,
    help="The SUMO configuration file for the scenario.")
```

File: rllib/examples/sumo_env_local.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+        "episode_reward_mean": args.stop_reward,
+    }
+
+    # Run the experiment.
+    results = tune.run("PPO",
+                       config=config,
+                       stop=stop,
+                       verbose=1,
+                       checkpoint_freq=10,
+                       restore=args.from_checkpoint)
```
Comment (sven1977) [inline rllib/examples/sumo_env_local.py:146]:
```suggestion
    results = tune.run(
        "PPO",
        config=config,
        stop=stop,
        verbose=1,
        checkpoint_freq=10,
        restore=args.from_checkpoint)
```

File: rllib/examples/sumo_env_local.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+logging.basicConfig(level=logging.WARN)
+logger = logging.getLogger("ppotrain")
+
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    "--sumo-connect-lib",
+    type=str,
+    default="libsumo",
+    choices=["libsumo", "traci"],
+    help="The SUMO connector to import. Requires the env variable SUMO_HOME set.")
```
Comment (sven1977) [inline rllib/examples/sumo_env_local.py:35]:
```suggestion
    help="The SUMO connector to import. "
    "Requires the env variable SUMO_HOME set.")
```

**Comment by LiCong2016** [comment]
**### I try to run the rllib/examples/sumo_env_local.py  code,  but  output error ** 

WARNING:tensorflow:From /root/anaconda3/envs/clever/lib/python3.7/site-packages/tensorflow/python/compat/v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.
Instructions for updating:
non-resource variables are not supported in the long term
2021-01-08 09:15:40,774	INFO services.py:1173 -- View the Ray dashboard at http://127.0.0.1:8265
Traceback (most recent call last):
  File "sumo_env_local.py", line 134, in <module>
    restore=args.from_checkpoint)
  File "/root/anaconda3/envs/clever/lib/python3.7/site-packages/ray/tune/tune.py", line 380, in run
    search_alg.add_configurations([exp])
  File "/root/anaconda3/envs/clever/lib/python3.7/site-packages/ray/tune/suggest/basic_variant.py", line 75, in add_configurations
    self._total_samples += count_variants(experiment.spec)
  File "/root/anaconda3/envs/clever/lib/python3.7/site-packages/ray/tune/suggest/variant_generator.py", line 142, in count_variants
    spec = copy.deepcopy(spec)
  File "/root/anaconda3/envs/clever/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/root/anaconda3/envs/clever/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/root/anaconda3/envs/clever/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/root/anaconda3/envs/clever/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/root/anaconda3/envs/clever/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/root/anaconda3/envs/clever/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/root/anaconda3/envs/clever/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/root/anaconda3/envs/clever/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/root/anaconda3/envs/clever/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/root/anaconda3/envs/clever/lib/python3.7/copy.py", line 220, in _deepcopy_tuple
    y = [deepcopy(a, memo) for a in x]
  File "/root/anaconda3/envs/clever/lib/python3.7/copy.py", line 220, in <listcomp>
    y = [deepcopy(a, memo) for a in x]
  File "/root/anaconda3/envs/clever/lib/python3.7/copy.py", line 169, in deepcopy
    rv = reductor(4)
TypeError: can't pickle module objects

**Comment by LiCong2016** [comment]
can you share more  about sumo demo code ?  Thank you very much!

**Pull request merged**